File: frontend/school-app/src/types/grade.ts
// types/grade.ts

export interface GradeResponse {
  id: number;
  studentId: number;
  studentName: string;
  subjectId: number;
  subjectName: string;
  classId: number;
  className: string;
  examName: string;
  examDate: string;          // ISO date string
  marksObtained: number;
  maxMarks: number;
  gradeLetter: string;
  remarks: string;
  enteredBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface GradeCreateRequest {
  studentId: number;
  subjectId: number;
  classId: number;
  examName: string;
  examDate: string;
  marksObtained: number;
  maxMarks: number;
  gradeLetter?: string;
  remarks?: string;
}

export interface GradeUpdateRequest {
  examName?: string;
  examDate?: string;
  marksObtained?: number;
  maxMarks?: number;
  gradeLetter?: string;
  remarks?: string;
}

// Used for filtering GET /grades
export interface GradeFilterParams {
  classId?: number;
  subjectId?: number;
  studentId?: number;
  className?: string | undefined;
}

File: schoolmanagementsystem/schoolmanagementsystem/src/main/java/com/school/schoolmanagementsystem/Entity/grade/Grade.java
package com.school.schoolmanagementsystem.Entity.grade;

import com.school.schoolmanagementsystem.Entity.Subject;
import com.school.schoolmanagementsystem.Entity.classes.ClassRoom;
import com.school.schoolmanagementsystem.Entity.users.Student;
import com.school.schoolmanagementsystem.Entity.users.Teacher;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "grades")
@Getter @Setter
@EntityListeners(AuditingEntityListener.class)
public class Grade {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "student_id", nullable = false)
    private Student student;

    @ManyToOne
    @JoinColumn(name = "subject_id", nullable = false)
    private Subject subject;

    @ManyToOne
    @JoinColumn(name = "class_id", nullable = false)
    private ClassRoom classRoom;

    @Column(nullable = false)
    private String examName;          // e.g., "Midterm", "Final Exam"

    private LocalDate examDate;

    @Column(nullable = false)
    private Double marksObtained;

    @Column(nullable = false)
    private Double maxMarks;

    private String gradeLetter;        // e.g., "A", "B+", computed or entered

    private String remarks;

    @ManyToOne
    @JoinColumn(name = "entered_by")
    private Teacher enteredBy;          // teacher who entered the grade

    @CreatedDate
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;
}
